### ISSUE TITLE
Reproducability with Seeds

### ISSUE THREAD

**Issue by engsbk** [lululxvi/deepxde#618]
Hello, @lululxvi, Thank you for you efforts!

Is it normal that I may get different prediction results even when using `dde.config.set_random_seed(12345)` in the beginning of my code? 

I'm including it with my imports but I still get different results of prediction. Is this the correct way to use it?
```
################################################################
########################SET SEED################################
################################################################
import deepxde as dde
from deepxde.backend import tf
import numpy as np
from scipy.interpolate import griddata
import matplotlib.pyplot as plt
import matplotlib
print(tf.__version__)
dde.config.set_random_seed(12345)

```

**Comment by lululxvi** [comment]
@pescap Any idea?

**Comment by lululxvi** [comment]
Fixed in the updated code.

**Comment by pescap** [comment]
I suggest that we open this issue again. I am working on a PR to fix the reproducibility issue for `tf.compat.v1` in GPU mode.

Reproducibility (tested with `Poisson_Lshape.py`).

| backend | GPU |  = | CPU |
| :---         |     :---:      |          :---: |  :---: |
| `tf.compat.v1`   |  x     |  x    | ✓
| `tensorflow`   |  ✓     | x     | ✓
| `pytorch`    | ✓       | ✓      |✓

✓: Reproducibility is ok
x: Reproducibility is not ok

Left column: Reproducibility in GPU mode
Center column: Results in GPU mode = Results in CPU mode 
Right column: Reproducibility in CPU modee

**Comment by lululxvi** [comment]
@pescap Poisson L-shape problem could be tricky, due to the L-BFGS.

**Comment by pescap** [comment]
Actually, I removed the L-BFGS lines. I will take the opportunity to test the behavior with the BFGS part.

**Comment by pescap** [comment]
Reproducibility (tested with `Poisson_Lshape.py`).

| backend | GPU |  = | CPU |
| :---         |     :---:      |          :---: |  :---: |
| `tf.compat.v1`   |  ✓     |  x    | ✓
| `tensorflow`   |  ✓     | x     | ✓
| `pytorch`    | ✓       | ✓      |✓

✓: Reproducibility is ok
x: Reproducibility is not ok

**Comment by jdellag** [comment]
> Reproducibility (tested with `Poisson_Lshape.py` over 5000 epochs) in #699.
> backend 	GPU 	= 	CPU
> `tf.compat.v1` 	✓ 	x 	✓
> `tensorflow` 	✓ 	x 	✓
> `pytorch` 	✓ 	✓ 	✓
> 
> ✓: Reproducibility is ok x: Reproducibility is not ok
> 
>     * No issues with L-BFGS except xx
> 
>     * For 'tensorflow', the results differ if you run twice the model in the same terminal. This behavior does not appear for 'tf.compat.v1'

I am not quite convinced of this for tf.compat.v1 backend. For example, here are two visualizations of my flow field for a Navier Stokes problem - ran with identical hyperparameters in separate Jupyter notebooks on the same GPU.
[Velocity_field_Re_0.01_beta_0_correct.pdf](https://github.com/lululxvi/deepxde/files/13686223/Velocity_field_Re_0.01_beta_0_correct.pdf)

[Velocity_field_Re_0.01_beta_0_incorrect.pdf](https://github.com/lululxvi/deepxde/files/13686226/Velocity_field_Re_0.01_beta_0_incorrect.pdf)

What could be the cause of such a difference? I have set the seed and my training points from both .dat files are identical.

**Comment by lululxvi** [comment]
@jdellag Could you design a very simple example to show what you found?

**Comment by jdellag** [comment]
Hi Lu,

I've poured a couple of hours here and there over the past few weeks into simplifying the problem so that I could reproduce these results but have not had any success. However, I am willing to share the entire code with you or any of the other developers so that perhaps we can pinpoint exactly where the discrepancy is and whether it is my code or the library. Please let me know how you'd like to proceed.

**Comment by pescap** [comment]
Just in case: For some backends, it is sometimes necessary to restart the Python kernel every time you run your simulation.

**Comment by jdellag** [comment]
While it did require a bit of work to move from a Jupyter setup to a batch job setup, I am happy to say that restarting the kernel for every simulation did indeed fix the issue. Thanks!!
